community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DiskLayer.java

```diff
@@ -179,70 +183,164 @@ public int labelGetForName( String label )
     }
 
     @Override
-    public boolean nodeHasLabel( long nodeId, int labelId ) throws EntityNotFoundException
+    public boolean nodeHasLabel( StoreStatement statement, long nodeId, int labelId ) throws EntityNotFoundException
     {
-        try
-        {
-            return PrimitiveIntCollections.indexOf( nodeGetLabels( nodeId ), labelId ) != -1;
-        }
-        catch ( InvalidRecordException e )
+        try ( StoreSingleNodeCursor nodeCursor = statement.acquireSingleNodeCursor() )
         {
-            return false;
+            if ( nodeCursor.init( nodeId ).next() )
+            {
+                try ( LabelCursor labelCursor = nodeCursor.get().labels() )
+                {
+                    return labelCursor.seek( labelId );
+                }
+            }
+            else
+            {
+                throw new EntityNotFoundException( EntityType.NODE, nodeId );
+            }
         }
     }
 
     @Override
-    public PrimitiveIntIterator nodeGetLabels( long nodeId ) throws EntityNotFoundException
+    public PrimitiveIntIterator nodeGetLabels( StoreStatement statement, long nodeId ) throws EntityNotFoundException
     {
-        NodeRecord record = nodeStore.loadRecord( nodeId, null );
-        if ( record == null )
-        {
-            throw new EntityNotFoundException( EntityType.NODE, nodeId );
-        }
-
-        final long[] labels = parseLabelsField( record ).get( nodeStore );
-        return new PrimitiveIntIterator()
+        try ( StoreSingleNodeCursor nodeCursor = statement.acquireSingleNodeCursor() )
         {
-            private int cursor;
-
 
-            @Override
-            public boolean hasNext()
+            if ( nodeCursor.init( nodeId ).next() )
             {
-                return cursor < labels.length;
-            }
+                final LabelCursor labelCursor = nodeCursor.get().labels();
 
+                return new PrimitiveIntIterator()
+                {
+                    boolean hasNext = labelCursor.next();
```

Comment:
True. Perhaps we can introduce specialized versions of `PrefetchingIterator` for `int` and `long`.